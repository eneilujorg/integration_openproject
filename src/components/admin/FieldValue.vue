<template>
	<div class="field-item">
		<b class="field-item-title">
			{{ t('integration_openproject', title) }}{{ isRequired ? '*' : '' }}:
		</b>

		<div class="field-item-value">
			{{ valueContent }}
		</div>

		<EyeIcon v-if="encryptValue && withInspection && !inspect"
			:size="16"
			class="field-item-inspect-btn"
			@click="toggleInspection" />
		<EyeOffIcon v-else-if="inspect"
			:size="16"
			class="field-item-inspect-off-btn"
			@click="toggleInspection" />
	</div>
</template>
<script>
import EyeIcon from 'vue-material-design-icons/Eye.vue'
import EyeOffIcon from 'vue-material-design-icons/EyeOff.vue'

export default {
	name: 'FieldValue',
	components: {
		EyeIcon, EyeOffIcon,
	},
	props: {
		title: {
			type: String,
			required: true,
		},
		value: {
			type: String,
			required: true,
		},
		isRequired: {
			type: Boolean,
			default: false,
		},
		encryptValue: {
			type: Boolean,
			default: false,
		},
		withInspection: {
			type: Boolean,
			default: false,
		},
	},
	data: () => ({
		inspect: false,
	}),
	computed: {
		encryptedValue() {
			return this.value.substring(0, 8) + '*'.repeat(15)
		},
		valueContent() {
			return (this.encryptValue && !this.inspect)
				? this.encryptedValue
				: this.value
		},
	},
	methods: {
		toggleInspection() {
			this.inspect = !this.inspect
		},
	},
}
</script>
<style lang="scss">
.field-item {
	padding: 6px 0;
	display: flex;
	align-items: center;
	&-value {
		padding: 0 4px;
	}
}
</style>
