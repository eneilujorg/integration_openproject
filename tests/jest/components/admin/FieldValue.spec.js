import { createLocalVue, mount } from '@vue/test-utils'
import FieldValue from '../../../../src/components/admin/FieldValue'

const localVue = createLocalVue()

const selectors = {
	eyeIcon: '.eye-icon',
	encryptedValue: '[data-test-id="encrypted-value"]',
}

describe('FieldValue', () => {
	describe('is required prop', () => {
		it('should append asterik with the title if set', () => {
			const wrapper = getWrapper({ isRequired: true })
			expect(wrapper).toMatchSnapshot()
		})
		it('should not append asterik with the title if not set', () => {
			const wrapper = getWrapper({ isRequired: false })
			expect(wrapper).toMatchSnapshot()
		})
	})
	describe('encrypt value prop', () => {
		describe("when set as 'true'", () => {
			it('should render the encrypted value', () => {
				const wrapper = getWrapper({
					encryptValue: true,
				})
				expect(wrapper.find(selectors.encryptedValue)).toMatchSnapshot()
			})
			describe('with inspection prop', () => {
				it('should show the eye icon if set', () => {
					const wrapper = getWrapper({
						encryptValue: true,
						withInspection: true,
					})
					expect(wrapper).toMatchSnapshot()
				})
				it('should render the toggler if both inspect and encrypt props are set', () => {
					const wrapper = getWrapper({
						encryptValue: true,
						withInspection: true,
					})
					expect(wrapper.find(selectors.eyeIcon).exists()).toBe(true)
				})
				it('should toggle encrypted value when eye icon is clicked', async () => {
					const wrapper = getWrapper({
						encryptValue: true,
						withInspection: true,
					})
					await wrapper.find(selectors.eyeIcon).trigger('click')
					expect(wrapper.find(selectors.encryptedValue)).toMatchSnapshot()
					await wrapper.find(selectors.eyeIcon).trigger('click')
					expect(wrapper.find(selectors.encryptedValue)).toMatchSnapshot()
				})
			})
		})
		it('should render the value as it is if not set', () => {
			const wrapper = getWrapper()
			expect(wrapper.find(selectors.encryptedValue).exists()).toBe(false)
		})
	})
})

function getWrapper(props = {}) {
	return mount(FieldValue, {
		localVue,
		propsData: {
			title: 'Some Field Title',
			value: 'Some Field Value',
			...props,
		},
		mocks: {
			t: (app, msg) => msg,
		},
	})
}
