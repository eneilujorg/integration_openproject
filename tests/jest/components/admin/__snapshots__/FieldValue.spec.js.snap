// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldValue encrypt value prop when set as 'true' should render the encrypted value 1`] = `<div data-test-id="encrypted-value"><span>Some Fie***************</span></div>`;

exports[`FieldValue encrypt value prop when set as 'true' with inspection prop should show the eye icon if set 1`] = `
<div class="field-value"><b class="title">
    Some Field Title:
  </b>
  &nbsp;
  <div data-test-id="encrypted-value"><span>Some Fie***************</span></div>
  <div class="eye-icon"></div>
</div>
`;

exports[`FieldValue encrypt value prop when set as 'true' with inspection prop should toggle encrypted value when eye icon is clicked 1`] = `<div data-test-id="encrypted-value"><span>Some Field Value</span></div>`;

exports[`FieldValue encrypt value prop when set as 'true' with inspection prop should toggle encrypted value when eye icon is clicked 2`] = `<div data-test-id="encrypted-value"><span>Some Fie***************</span></div>`;

exports[`FieldValue is required prop should append asterik with the title if set 1`] = `
<div class="field-value"><b class="title">
    Some Field Title*:
  </b>
  &nbsp;
  <span>Some Field Value</span>
  <!---->
</div>
`;

exports[`FieldValue is required prop should not append asterik with the title if not set 1`] = `
<div class="field-value"><b class="title">
    Some Field Title:
  </b>
  &nbsp;
  <span>Some Field Value</span>
  <!---->
</div>
`;
